crates/astria-conductor/src/conductor.rs

```diff
@@ -10,122 +18,239 @@ use tokio::{
+            data_availability_reader = Some(reader);
+        };
+
         Ok(Self {
             signals,
-            driver,
-            executor_join_handle,
-            reader_join_handle,
+            tasks,
+            sequencer_client_pool,
+            shutdown_channels,
+            sync_done,
+            data_availability_reader,
         })
     }
 
     pub async fn run_until_stopped(self) -> eyre::Result<()> {
+        use futures::future::{
+            FusedFuture as _,
+            FutureExt as _,
+        };
+
         let Self {
             signals:
                 SignalReceiver {
                     mut reload_rx,
                     mut stop_rx,
                 },
-            driver,
-            executor_join_handle: mut executor,
-            reader_join_handle: mut reader,
+            mut tasks,
+            shutdown_channels,
+            sequencer_client_pool,
+            sync_done,
+            mut data_availability_reader,
         } = self;
-        let driver_tx = driver.cmd_tx.clone();
-        let mut driver = tokio::spawn(driver.run());
-        let mut interval = time::interval(Duration::from_secs(3));
+
+        let mut sync_done = sync_done.fuse();
 
         loop {
             select! {
-                // FIXME: The bias should only be on the signal channels. The the two
-                //       handlers should have the same bias.
+                // FIXME: The bias should only be on the signal channels. The two handlers should have the same bias.
                 biased;
 
                 _ = stop_rx.changed() => {
                     info!("shutting down conductor");
-                    if let Some(e) = driver_tx.send(DriverCommand::Shutdown).err() {
-                        error!(
-                            error.msg = %e,
-                            error.cause = ?e,
-                            "error sending Shutdown command to driver"
-                        );
-                    }
                     break;
                 }
 
                 _ = reload_rx.changed() => {
                     info!("reloading is currently not implemented");
                 }
-                // request new blocks every X seconds
-                _ = interval.tick() => {
-                    if let Some(e) = driver_tx.send(DriverCommand::GetNewBlocks).err() {
-                        // the only error that can happen here is SendError which occurs
-                        // if the driver's receiver channel is dropped
-                        error!(
-                            error.msg = %e,
-                            error.cause = ?e,
-                            "error sending GetNewBlocks command to driver"
-                        );
-                        break;
-                    }
-                }
 
-                res = &mut driver => {
+                res = &mut sync_done, if !sync_done.is_terminated() => {
```

Comment:
would there be a case where `sync_done` is ready but `sync_done.is_terminated()`? reading the docs, the future is supposed to be terminated when it returns `Poll::Ready` right? https://docs.rs/futures/latest/futures/future/trait.FusedFuture.html

Reply:
The if-block is evaluated before the future is polled. So the arm is only deactivated if and only if there are no more futures in the stream.